packages/eslint-plugin/src/rules/no-floating-promises.ts

```diff
@@ -325,20 +373,46 @@ export default createRule<Options, MessageId>({
   },
 });
 
-function isPromiseArray(checker: ts.TypeChecker, node: ts.Node): boolean {
+function doesTypeMatchSpecifier(
```

Comment:
I dont really understand the point being made here though.

Are you saying the rule should report in this case ? [Playground link line 4](https://deploy-preview-8502--typescript-eslint.netlify.app/play#ts=5.4.3&fileType=.tsx&code=MYGwhgzhAEDKYDMCmAFATgewLYEsJIB4AVAPmiQA8AXJAOwBMZ1s9DToBvaAXwFgAoYBloQq0CIlSZc%2BAFxxJzGYVoBXLACMkaMgF5oS1gDo0SCBhAA3JAAoArAEoA3AID0AKnIQQOWmIBEAAJUAJ4ADmbAaDhhVAC0Zj5%2BrrQYcQggGGBUvgDmcWHSrBD%2B8gDa-tqYaP4ANJzQYCCZAO4AYhhoANKpLbTwyIb4EOVcCNLyAOQIOCBIk-W0YFhIUwNSLPiTPAC6PNB77q4CEoNF%2BEbA2cAAFjbOAkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

& the reason is - `catch()` returns a normal `Promise` ??

Reply:
Hmmm. I think I see your point.... but let's drill down into the details a bit, here. There's a difference between `.catch()` and `.catch(() => {})`. 

for normal promises,
```ts
// doesn't flag, promise is handled
normalPromise.catch(() => {});
// does flag! catch requires a (function) argument to handle the promise
normalPromise.catch();
```

so, my thought would be
```ts
// shouldn't flag; since this would be considered a safe way to handle any thenable
// (hopefully won't be encountered in real world code, anyway, since the whole point
// is that you shouldn't need to .catch a safe promise)
safePromise.catch(() => {});
// likewise, shouldn't flag
safePromise.then(() => {}, () => 'has rejection handler');

// should flag - this wouldn't be a valid way of handling a rejection in a normal promise,
// and we don't know what the safePromise .catch does other than that it creates a normal promise 
// that's unhandled 🤷 
safePromise.catch()
```
But, I don't feel certain that my interpretation here is unambiguously correct, so feel free to make the case that it should go the other way 🙂 

Regardless of the above, I think that this test case unambiguously needs to flag (provided safePromise produces a normal Promise, not another SafePromise). 
```ts
safePromise.then(() => { throw new Error(); });
```

Reply:
> other than that it creates a normal promise

It isn't ? Lets take this example
```ts
class SafePromise<T> extends Promise<T> { }
const safePromise: SafePromise<number> = Promise.resolve(5);
safePromise.catch(); // the type of safePromise is `SafePromise<number>` hover on the variable name in vscode
```

> we don't know what the safePromise .catch does

Yeah, this is kindof what this PR is trying to do.

The `allowForKnownSafePromises` option is letting the rule know which types are safe to-not-handle-rejection.
To-not-handle-rejection means `.catch` can have 0 arguments, `.then` can have 1 arg etc ...

Lets take an example
```ts
class SafePromise<T> extends Promise<T> { }
const safePromise: SafePromise<number> = Promise.resolve(5);
/* eslint "@typescript-eslint/no-floating-promises": ["error", { allowForKnownSafePromises: [{ from: 'file', name: 'SafePromise' }] } ] */
safePromise.catch();
```
1. Here the rule finds out what the type of `safePromise` is and checks if it matches with any type present in the config file/comment. 
2. If it matches, the rule says it is safe (not a floating promise & doesn't report), else(if nothing is matched), it continues checking whether `.catch` has 1 arg or `.then` has 2 etc...

... since `safePromise.catch()` is a `CallExpression` & its callee's type is `MemberExpression`, thats why the `if` is located [here](https://github.com/typescript-eslint/typescript-eslint/pull/8502/files#diff-dec933bc2964d8570e62f3f2cd08e4237ee8e2dcfe894690afa5042095385089R285). & since there are so many `ExpressionStatement`s thats why the diff contains so many `if`s

Reply:
I see what you're saying.... but I think we're talking past each other a tiny bit on the details.

Bringing it back to the simpler and more impactful corollary - shouldn't this flag? (current version of code doesn't)
```ts
class SafePromise<T> extends Promise<T> { }
const safePromise = new SafePromise<void>((resolve, reject) => { resolve() });
safePromise.then(() => { throw new Error(); });
```

`safePromise` is a `SafePromise`, of course, but the _whole_ expression is an _ordinary_ Promise, due to `.then` still returning a new ordinary Promise.

I get that `SafePromise` is "safe to-not-handle-rejection". But that wouldn't imply that it's safe to chain an unhandled ordinary `Promise` off of, right? From reading the issue and comments, I don't see somewhere that this part is spelled out explicitly (feel free to point me to it if it has already been discussed). But my current interpretation is that it should flag (unless, of course, `SafePromise` is modified to also override the `.then` to return another `SafePromise`).

Thoughts?

(The handlerless `.catch()` is a more subtle and less important version of the above, so my thought is to get to the same page about this first)

Reply:
Thanks for the comments, I did understood, tried to implement it locally & here are my few thoughts.

First, I dont think this rule ever checked whether the promises returned by `then`/`catch` are floating or not(with or without this PR)
[Minimum repro](https://typescript-eslint.io/play/#ts=5.4.3&fileType=.tsx&code=AoJw9gtglgzgpgOhHAVnAxgFwBQDYCUC6AhpugBbZwggAEAvAHy0DeAsAFC22bngDutAHZxBAURpgQVGvgDcnAL75aAelU9ycWgAdw0eLWSYAriBEATWgCMAnrQAGJMuQe1YtAOZxMmKEM8jVAxMOCtiISsQEwAbbQswOBghAHJMIJ0pdIjbXn9PTk5QSFhEZDQsPEJeOCFsAA8GZnqAGlpqOiZWTm5eAWFRWglwaQ75JRV1TW09EsNjM0sbewcaoTcPb198oIrQ8MijWPjE5LSMrNocvICgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDN4OBDfMwDmtYtA4BbSshTooiaBOiRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)
In both the cases, the promises returned by `then`/`catch` are getting rejected(run the code in browser console, it'll show `Uncaught ( in promise )`). So ... if the rule starts unflagging promises returned by `then` with 2 args(in your comment above), it might be a false positive(since they can also be floating).

Second, is it even possible for the rule to determine whether the promises returned `then`/`catch` are floating or not ? :sweat_smile:. It depends on purely on the `onRejected` function(not on the absence or presence of it). The 1st code snippet [here](https://github.com/typescript-eslint/typescript-eslint/pull/8502#discussion_r1586373572) is wrong ?

So, having a function as an argument in `catch`/`then` doesnt ensure that the promise returned by it will not get rejected(same as the 1st one). & I dont think this rule can ever decide whether the promise returned by it will get rejected or not(maybe :sweat_smile:) 

(Because its not only `throw` statements, anything can make it reject)
![catch's promise rejection](https://github.com/typescript-eslint/typescript-eslint/assets/105232141/87e2edce-a4a0-4175-9f24-f9c6728def32)

Reply:
I'm talking about `.then` _without_ a rejection handler.
The rule flags this:
```ts
promise.then(f);
```
_regardless_ of what `f` is. The question is whether it should flag this
```ts
safePromise.then(f);
```

Reply:
No, it shouldn't, 

Because unflagging can give a false positive.

We're talking about that promise which is returned by `then` right ? Not the promise on which `then` is attached.

Reply:
> Because unflagging can give a false positive.

I'm not sure I follow... Could you explain what this means?

> We're talking about that promise which is returned by `then` right ? Not the promise on which `then` is attached.(both are different ones)

I'm saying this:
```ts
// (no error)
safePromise;

// error - expression evaluates to an ordinary unhandled promise
safePromise.then(f);
~~~~~~~~~~~~~~~~~~~
```

Reply:
This expression will still evaluate to an ordinary unhandled promise ( which will reject )
```ts
Promise.reject(1).then(x => x, err => {
  throw new Error(2);
});
```

Reply:
That is a true statement (and if you're interested in the reasoning behind that design decision, you may be interested to read up starting from here: https://github.com/typescript-eslint/typescript-eslint/issues/5692#issuecomment-1504645034). 

However, that doesn't really impact this PR. A SafePromise should basically be granted exactly _one_ extra privilege over what the rule currently allows, and that is that a chain ending in a `SafePromise` (or, I suppose, `Array<SafePromise>`) should be ignored. Otherwise, a SafePromise should function the same as a Promise.

```ts
// unhandled
promise;
// handled!
safePromise;

// unhandled
[promise, promise, promise];
// handled
[safePromise, safePromise, safePromise];
// mixed should be unhandled
[safePromise, promise, safePromise];

// both unhandled (important!)
promise.then(f);
safePromise.then(f);

// both handled
promise.catch(f);
safePromise.catch(f);

// both handled, intentional design decision
promise.then(f, g);
safePromise.then(f, g);

// both unhandled 
promise.catch();
safePromise.catch();

// both unhandled (silly edge case)
promise.then(f, undefined);
safePromise.then(f, undefined);

// finally is basically chopped off the end for the purposes of this rule, so
// unhandled
promise.finally(f);
// handled
safePromise.finally(f);
```

Would you kindly update the PR with test cases for these (modifying the existing test cases that contradict these)? I think you'll find that it will also result in a much simpler implementation for the rule changes, as mentioned in the first post in this thread. 

Thanks!